facebook/OpenBIC
# fby35: cl: Support read DIMM via I3C

Summary:
- Monitor DIMM's information (DIMM's temperature, PMIC power, and PMIC error) through I3c when DIMMs are present.

Test plan:
- Build code: Pass
- Sensor of DIMM related reading: Pass
- Monitor PMIC error status: Pass

Log:
1. Check DIMM related sensor reading is normal. root@bmc-oob:~# sensor-util slot1 --force | grep DIMM
MB_DIMMA0_TEMP_C             (0x6) :   41.50 C     | (ok)
MB_DIMMA2_TEMP_C             (0x7) :   40.50 C     | (ok)
MB_DIMMA3_TEMP_C             (0x9) :   39.00 C     | (ok)
MB_DIMMA4_TEMP_C             (0xA) :   40.25 C     | (ok)
MB_DIMMA6_TEMP_C             (0xB) :   37.75 C     | (ok)
MB_DIMMA7_TEMP_C             (0xC) :   35.75 C     | (ok)
MB_ADC_P12V_DIMM_VOLT_V      (0x26) :   12.63 Volts  | (ok)
MB_VR_DIMMA0_PMIC_PWR_W      (0x1E) :    0.50 Watts  | (ok)
MB_VR_DIMMA2_PMIC_PWR_W      (0x1F) :    0.62 Watts  | (ok)
MB_VR_DIMMA3_PMIC_PWR_W      (0x36) :    0.62 Watts  | (ok)
MB_VR_DIMMA4_PMIC_PWR_W      (0x37) :    0.38 Watts  | (ok)
MB_VR_DIMMA6_PMIC_PWR_W      (0x42) :    0.38 Watts  | (ok)
MB_VR_DIMMA7_PMIC_PWR_W      (0x47) :    0.75 Watts  | (ok)

2. Using raw command to inject PMIC critical error and check error SEL.
- Inject DIMM critical error SWAout OV on DIMM A4. [BIC console]
uart:~$ i3c xfer I3C_3 -a 0x4c -w 0x35,0x90 -r 1
Private transfer to address 0x4c

00Send gpio interrupt to BMC, gpio number(32) status(0) 000000: 9Send gpio interrupt to BMC, gpio number(39) status(0)
0                                               |.                |
uart:~$ [00:01:34.528,000] <wrn> power_status: POST_COMPLETE: no
[00:01:34.529,000] <wrn> power_status: DC_STATUS: off
[00:01:34.530,000] <wrn> power_status: CPU_PWR_GOOD: no
[00:01:34.528,000] <wrn> power_status: POST_COMPLETE: no
[00:01:34.529,000] <wrn> power_status: DC_STATUS: off
[00:01:34.530,000] <wrn> power_status: CPU_PWR_GOOD: no
uart:~$ slp3

- Check PMIC error SEL. [BMC console]
root@bmc-oob:~# log-util --print slot1
2023 Jan 04 22:18:01 log-util: User cleared FRU: 1 logs
1    slot1    2023-01-04 22:19:13    ipmid            SEL Entry: FRU: 1, Record: Standard (0x02), Time: 2023-01-04 22:19:13, Sensor: PMIC_ERR (0xB4), Event Data: (030000) DIMM A4 SWAout OV Assertion
1    slot1    2023-01-04 22:19:18    ipmid            SEL Entry: FRU: 1, Record: Standard (0x02), Time: 2023-01-04 22:19:18, Sensor: ME_POWER_STATE (0x16), Event Data: (000000) RUNNING Assertion
1    slot1    2023-01-04 22:19:27    ipmid            SEL Entry: FRU: 1, Record: Standard (0x02), Time: 2023-01-04 22:19:27, Sensor: SYSTEM_STATUS (0x10), Event Data: (0AFFFF) S0 power on failure Assertion

3. Using raw command to inject PMIC warning error and check error SEL.
- Inject DIMM warning error high_temp on DIMM A4. [BIC console]
uart:~$ i3c xfer I3C_3 -a 0x4c -w 0x35,0x83 -r 1
Private transfer to address 0x4c

00000000: 83

- Check PMIC error SEL. [BMC console]
root@bmc-oob:~# log-util --print slot1
2023 Jan 04 22:18:01 log-util: User cleared FRU: 1 logs
1    slot1    2023-01-04 22:18:59    ipmid            SEL Entry: FRU: 1, Record: Standard (0x02), Time: 2023-01-04 22:18:59, Sensor: PMIC_ERR (0xB4), Event Data: (030C00) DIMM A4 High temp warning Assertion
